#HELLO WORLD!

**Instructions:**


1.   Submit both .ipynb and PDF files, No pdf or no .ipynb will get '0'.
2.   Do not ZIP the files and submit, Zip submission will get '0'.
3.   Provide Observations for all questions.
4.   **Similarity score must be less than 15%**
5.   Don't use same datasets used in tutorials for tasks.
6.   click here for datasets used in tutorials [click here](https://drive.google.com/file/d/1ZT3hnZHeoCht-URF3zPND4JMMz9CGXWb/view?usp=sharing)

**RUBRIC:**


**TASK 1 :** **10pts full pts | 0pts No pts**

**TASK 2 :** **10pts full pts | 0pts No Ans**

**Task 3 :** **25pts Full Marks | 0pts No Answer**

**TASK 4 :**   **30pts Full marks | 0pts No Answer**

**TASK 5 :** **10pts full marks | 0pts No Answer**

**TASK 6 :** **10pts full marks | 0pts No Ans**

**TASK 7 :** **5pts full marks | 0pts No Ans**


#**Spelling Correction in Natural Language Processing**



Please follow the tutorials below and complete the tasks that are available on the webpages provided. The tutorials will have code and might not have dataset file. You can create dataset files as per the tutorial requirements. The aim of the ICE is to make the tutorials in executed format. You can also use the github repositories of the authors if they are avilable. It is recommended to run the tutorials and then test it with your own datasets (Custom made). You can use any source on the internet to complete the tasks. For task 4 your have to provide mini-examples to differentiate between non-word and real world spelling corrections. After the code for task 4 please provide a brief explanation for what is the difference and what is your analysis. 

# Tutorial 1:
### Use the following tutorial to implement spelling checking using Textblob


In [69]:
!pip install textblob

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


**The correct() Function**

In [70]:
from textblob import TextBlob

with open("text.txt", "r") as f:        # Opening the test file with the intention to read
    text = f.read()                     # Reading the file
    textBlb = TextBlob(text)            # Making our first textblob
    textCorrected = textBlb.correct()   # Correcting the text
    print(textCorrected)

Of we admit the the geological recording is imperfect in an extreme degree, than such facts as the record gives, support the theory of descent with modification. New species hive come on the stage slowly and at successive intervals; and the amount of change, after equal intervals of time, is widely different in different groups.


###**How Correct is TextBlob's Spelling Correction?**
As we can see, the text still has some spelling errors. Words like "abl" were supposed to be "able", not "all". Though, even with these, it's still better than the original.

**Now comes the question, how much better is it?**


###The following code snippet is a simple script that test how good is TextBlob in correcting errors, based on this example:

In [71]:
from textblob import TextBlob

# A function that compares two texts and returns 
# the number of matches and differences
def compare(text1, text2):  
    l1 = text1.split()
    l2 = text2.split()
    good = 0
    bad = 0
    for i in range(0, len(l1)):
        if l1[i] != l2[i]:
            bad += 1
        else:
            good += 1
    return (good, bad)

# Helper function to calculate the percentage of misspelled words
def percentageOfBad(x):
    return (x[1] / (x[0] + x[1])) * 100

In [72]:
with open("test.txt", "r") as f1: # test.txt contains the same typo-filled text from the last example 
    t1 = f1.read()

with open("original.txt", "r") as f2: # original.txt contains the text from the actual book 
    t2 = f2.read()

t3 = TextBlob(t1).correct()

mistakesCompOriginal = compare(t1, t2)
originalCompCorrected = compare(t2, t3)
mistakesCompCorrected = compare(t1, t3)

print("Mistakes compared to original ", mistakesCompOriginal)
print("Original compared to corrected ", originalCompCorrected)
print("Mistakes compared to corrected ", mistakesCompCorrected, "\n")

print("Percentage of mistakes in the test: ", percentageOfBad(mistakesCompOriginal), "%")
print("Percentage of mistakes in the corrected: ", percentageOfBad(originalCompCorrected), "%")
print("Percentage of fixed mistakes: ", percentageOfBad(mistakesCompCorrected), "%", "\n")

Mistakes compared to original  (43, 12)
Original compared to corrected  (50, 5)
Mistakes compared to corrected  (44, 11) 

Percentage of mistakes in the test:  21.818181818181817 %
Percentage of mistakes in the corrected:  9.090909090909092 %
Percentage of fixed mistakes:  20.0 % 



###As we can see, the correct method managed to get our spelling mistake percentage from 20% to 9%, which is pretty decent, however there's a bit of a catch. It corrected 20% of the words, so why is there still a 0.59% mistake rate?

The answer is overcorrection. Sometimes, it can change a word that is spelled correctly, like the first word in our example text where "As" was corrected to "Is". Other times, it just doesn't have enough information about the word and the context to tell which word the user was intending to type, so it guesses that it should replace "whl" with "while" instead of "whole".

There is no perfect spelling corrector because so much of spoken language is contextual, so keep that in mind. In most use cases, there are way fewer mistakes than in our example, so TextBlob should be able to work well enough for the average user.

# Task 1:
### Use the above tutorial to implement spelling checking using Textblob on your own dataset.
###write your own observation on the functioning

# Tutorial 2: 
###Training TextBlob with Custom Datasets

In [73]:
from textblob.en import Spelling        
import re

textToLower = ""

with open("originOfSpecies.txt","r") as f1:           # Open our source file
	text = f1.read()                                  # Read the file                 
	textToLower = text.lower()                        # Lower all the capital letters

words = re.findall("[a-z]+", textToLower)             # Find all the words and place them into a list    
oneString = " ".join(words)                           # Join them into one string

pathToFile = "train.txt"                              # The path we want to store our stats file at
spelling = Spelling(path = pathToFile)                # Connect the path to the Spelling object
spelling.train(oneString, pathToFile)                 # Train

In [74]:
from textblob.en import Spelling        
from textblob import TextBlob

pathToFile = "train.txt" 
spelling = Spelling(path = pathToFile)
text = " "

with open("test.txt", "r") as f: 
	text = f.read()

words = text.split()
corrected = " "
for i in words :
    corrected = corrected +" "+ spelling.suggest(i)[0][0] # Spell checking word by word

print(corrected)

  Of we admit the the geological record is imperfect in an extreme degree than such facts as the record gives support the theory of descent with modification New species hive come on the stage slowly and at successive intervals and the amount of change after equal intervals of time is widely different in different groups


###This fixes around 2 out of 3 of misspelled words, which is pretty good, considering the run without much context.

#**Conclusion**
Here used TextBlob to implement a basic spelling corrector, both with the stock prediction model a custom one.

Correcting man-made spelling errors has become a common task for software developers. Even though it has become easier and more efficient via data mining, many spelling mistakes need context to be corrected.

In conclusion, proofreaders are probably not going to get automated out of work any time soon, though, some basic correction can be automated to save time and effort.

# Task 2: 
### Train your Textblob model on your own dataset
### Instructions are provided in the above *tutorial*

###write your own observation on the functioning

# Tutorial 3: 
### Implementation Petr Norvig algorithm for spelling corrections. 


In [75]:
import re
from collections import Counter

def words(text): return re.findall(r'\w+', text.lower())

WORDS = Counter(words(open('/content/big.txt').read()))

def P(word, N=sum(WORDS.values())): 
    "Probability of `word`."
    return WORDS[word] / N

def correction(word): 
    "Most probable spelling correction for word."
    return max(candidates(word), key=P)

def candidates(word): 
    "Generate possible spelling corrections for word."
    return (known([word]) or known(edits1(word)) or known(edits2(word)) or [word])

def known(words): 
    "The subset of `words` that appear in the dictionary of WORDS."
    return set(w for w in words if w in WORDS)

def edits1(word):
    "All edits that are one edit away from `word`."
    letters    = 'abcdefghijklmnopqrstuvwxyz'
    splits     = [(word[:i], word[i:])    for i in range(len(word) + 1)]
    deletes    = [L + R[1:]               for L, R in splits if R]
    transposes = [L + R[1] + R[0] + R[2:] for L, R in splits if len(R)>1]
    replaces   = [L + c + R[1:]           for L, R in splits if R for c in letters]
    inserts    = [L + c + R               for L, R in splits for c in letters]
    return set(deletes + transposes + replaces + inserts)

def edits2(word): 
    "All edits that are two edits away from `word`."
    return (e2 for e1 in edits1(word) for e2 in edits1(e1))

################ Test Code 

def unit_tests():
    assert correction('speling') == 'spelling'              # insert
    assert correction('korrectud') == 'corrected'           # replace 2
    assert correction('bycycle') == 'bicycle'               # replace
    assert correction('inconvient') == 'inconvenient'       # insert 2
    assert correction('arrainged') == 'arranged'            # delete
    assert correction('peotry') =='poetry'                  # transpose
    assert correction('peotryy') =='poetry'                 # transpose + delete
    assert correction('word') == 'word'                     # known
    assert correction('quintessential') == 'quintessential' # unknown
    assert words('This is a TEST.') == ['this', 'is', 'a', 'test']
    assert Counter(words('This is a test. 123; A TEST this is.')) == (
           Counter({'123': 1, 'a': 2, 'is': 2, 'test': 2, 'this': 2}))
    assert len(WORDS) == 32198
    assert sum(WORDS.values()) == 1115585
    assert WORDS.most_common(10) == [
 ('the', 79809),
 ('of', 40024),
 ('and', 38312),
 ('to', 28765),
 ('in', 22023),
 ('a', 21124),
 ('that', 12512),
 ('he', 12401),
 ('was', 11410),
 ('it', 10681)]
    assert WORDS['the'] == 79809
    assert P('quintessential') == 0
    assert 0.07 < P('the') < 0.08
    return 'unit_tests pass'

def spelltest(tests, verbose=False):
    "Run correction(wrong) on all (right, wrong) pairs; report results."
    import time
    start = time.clock()
    good, unknown = 0, 0
    n = len(tests)
    for right, wrong in tests:
        w = correction(wrong)
        good += (w == right)
        if w != right:
            unknown += (right not in WORDS)
            if verbose:
                print('correction({}) => {} ({}); expected {} ({})'
                      .format(wrong, w, WORDS[w], right, WORDS[right]))
    dt = time.clock() - start
    print('{:.0%} of {} correct ({:.0%} unknown) at {:.0f} words per second '
          .format(good / n, n, unknown / n, n / dt))
    
def Testset(lines):
    "Parse 'right: wrong1 wrong2' lines into [('right', 'wrong1'), ('right', 'wrong2')] pairs."
    return [(right, wrong)
            for (right, wrongs) in (line.split(':') for line in lines)
            for wrong in wrongs.split()]

if __name__ == '__main__':
    print(unit_tests())
    spelltest(Testset(open('/content/spell-testset1.txt')))
    spelltest(Testset(open('/content/spell-testset2.txt')))

unit_tests pass


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:74: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:85: DeprecationWarning: time.clock has been deprecated in Python 3.3 and will be removed from Python 3.8: use time.perf_counter or time.process_time instead


75% of 270 correct (6% unknown) at 41 words per second 
68% of 400 correct (11% unknown) at 46 words per second 


In [76]:
correction('juse')

'just'

###So on the development set we get 75% correct (processing words at a rate of 41 words/second), and on the final test set we get 68% correct (at 46 words/second). In conclusion, I met my goals for brevity, development time, and runtime speed, but not for accuracy. Perhaps my test set was extra tough, or perhaps my simple model is just not good enough to get to 80% or 90% accuracy.

# Task 3: 
### Perform Petr Norvig algorithm for spelling corrections  using other dataset.
###Also provide your observation for this algorithm


# Task 4: (40%)
### Implement spelling correction using Noisy Channel for non-word and real word. You can follow any code

###provide your observation

#Task 5:

###Implement spelling correction using edit distance model.

#Task:6

###Implement spelling correction using Spacy

#Task 7:
###Explain your observation for spelling correction using edit distance and spacy. which one do you think is better?

###REFERENCES
1) 
https://stackabuse.com/spelling-correction-in-python-with-textblob/

2) 
https://medium.com/mlearning-ai/build-spell-checking-models-for-any-language-in-python-aa4489df0a5f